### PR TITLE
8604 improve volta unwinding

### PR DESCRIPTION
**Pull request by jeetee**
no file format changes required, so should quite cleanly apply to 2.x as well; but does change volta interpretation.
Where 2.x currently requires the repeatlist of a volta to be set to the n-th time that measure/volta is encountered, this PR requires the repeatlist to reflect the human-readable logic of being the x-th repeat in total.

Applying this to 2.x will result in scores pre 2.2 with volta rendering a different playback roadmap than scores from 2.2+

### CODE REVIEW

**Comment by Jojo-Schmitz** [comment]
As it never worked as intended prior to this PR, I tend to think that we should fix it for 2.2 too, even if that means that playback may render differently, due to the tricks involved getting it to work half way decent in pre 2.2. Together with a sentence about 'Known incompatibilities' in the Release Notes of course...

File: libmscore/repeatlist.cpp [hunk 1]
```diff
@@ -30,7 +30,7 @@ Volta* Score::searchVolta(int tick) const
       {
       for (const std::pair<int,Spanner*>& p : _spanner.map()) {
             Spanner* s = p.second;
-            if (s->type() != ElementType::VOLTA)
+            if (!s->isVolta())
```
Comment (Jojo-Schmitz) [inline libmscore/repeatlist.cpp]: I believe this and some more `isXxx()` and `toXxx()` won't fit for 2.2.

Comment (jeetee) [reply libmscore/repeatlist.cpp]: indeed.. if desired by @lasconic I could rework this into two separate commits, then the first one would remain 2.x compatible and all "codestyle" changes into a 2nd commit?

Comment (lasconic) [reply libmscore/repeatlist.cpp]: I can handle this. Give me a couple of days to try the PR and see how much it improve  the current situation and break compatibility.

**Comment by MarcSabatella** [comment]
FWIW, I definitely applaud any improvement that can be made here, and would be willing to bear the pain of some incompatibility if it means we can finally have a more intuitive and logical story to tell.  Right now, of all the FAQ's we get, it is the questions about volta behavior that bother me the most, because we just don't *have* great answers.

However, completely breaking compatibility - so that even the most basic tunes with a single repeat and two voltas no longer works - would make me uncomfortable.  I'd like to understand better just how bad it will be.

**Comment by jeetee** [comment]
It will 'break' the most simple volta song from 2.x if the volta properties for it had been altered to achieve correct playback in 2.x

If you simply had consecutive voltas, 2.x required you to set the repeat lists as follows:
volta_flaws_in_v2
With this PR, only the first one of those voltas will be played back, and the other two will be skipped entirely. The now 'correct' way to set the repeat lists is as follows:
volta_flaws_corrected_in_pr

The bonus is that this currently is how the repeat lists are set for the palette items in 2.x already. So a user would in the most basic scenarios no longer be required to fiddle with the repeatlist property at all.

**Comment by Jojo-Schmitz** [comment]
So this PR will break the playback of 2.x scores that used a workaround to get the playback right prior to this fix. IMHO this is a downside we can quite easily accept

File: mtest/libmscore/repeat/repeat13.mscx [hunk 2]
```diff
@@ -112,14 +112,6 @@
           </BarLine>
         </Measure>
       <Measure number="5">
-        <Jump>
```
Comment (lasconic) [inline mtest/libmscore/repeat/repeat13.mscx]: I don't get that. Why is it removed?

**Comment by lasconic** [comment]
Except my comment on test repeat13.mscx, it works great. I'm checking the code, and what needs to be done for 2.2.

File: mtest/libmscore/repeat/repeat13.mscx (hunk 2 above)
Comment (lasconic) [reply mtest/libmscore/repeat/repeat13.mscx]: Ok now I see the commit message "no-repeat (13) shouldn't have an off-page invalid jump either". In fact, this test was there to make sure that meaningless jump would do nothing. It should stay like this, and indeed the jump shouldn't be offpage.

File: libmscore/repeatlist.cpp [hunk 3]
```diff
@@ -327,180 +301,235 @@ void RepeatList::unwind()
...
-            if (m->repeatStart())
-                  break;
-
-            if (m == _score->firstMeasure())
-                  break;
-
-            if (m->prevMeasure()->sectionBreak())
-                  break;
-
+      while (   (!m->repeatStart())
```
Comment (lasconic) [inline libmscore/repeatlist.cpp]: Too much spaces here.

File: mtest/libmscore/repeat/tst_repeat.cpp [hunk 4]
```diff
@@ -83,6 +83,17 @@ class TestRepeat : public QObject, public MTest
...
+      void repeat39() { repeat("repeat39.mscx", "1;2;3; 2;4;5; 2;3; 2;4;5; 2;6"); } // volta 1.3.,2.4.
+
+      void repeat40() { repeat("repeat40.mscx", "1;2;1;3;4;1;5"); } // #148276 Open volta lines must extend over entire ending: next endRepeat
+      void repeat41() { repeat("repeat41.mscx", "1;2;1;3;1;4;5"); } // #148276 Open volta lines must extend over entire ending: next endRepeat at volta (SLine) end
+      void repeat42() { repeat("repeat42.mscx", "1;2;1;3;5;1;4;5"); } // #148276 Open volta lines must extend over entire ending: next volta
+      void repeat43() { repeat("repeat43.mscx", "1;2;1;2;1;3;1;2;1;4;1;5"); } // #8604 Complex alternate endings 1.2.4. / 3./ 5.
+      void repeat44() { repeat("repeat44.mscx", "1;2;3;4;5;3;6;3;4;1;7"); } // Jump from within a volta
+      
+      void repeat45() { repeat("repeat45.mscx", "1;2;3;4;3;5;6;2;3;4;3;5;6;7"); } // repeat12 but with 'play repeats' enabled
+	  void repeat46() { repeat("repeat46.mscx", "1;2;3;4;5;3;4;2;3;4;5;3;4;5;6"); } // repeat24 but with 'play repeats' enabled
```
Comment (lasconic) [inline mtest/libmscore/repeat/tst_repeat.cpp]: too much spaces here too.

**Comment by lasconic** [comment]
I fixed the issues mentioned earlier and merged the PR manually at 30879209f5c601eafbfe96eaaa1605359d718a25 and dd2050e2f1d820570e0f7e05e7fec51a3a6380ec. 
I will create a PR for 2.2 but I might need some help in order to create  the missing test files in 2.x.

**Comment by lasconic** [comment]
Here is the PR https://github.com/musescore/MuseScore/pull/3442

**Pull request closed without merging**